File: backend/domain/src/main/java/io/dkakunsi/bitapp/loan/model/Loan.java
package io.dkakunsi.bitapp.loan.model;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.Currency;

import io.dkakunsi.bitapp.common.ModelStatus;
import io.dkakunsi.bitapp.user.model.User;

public final record Loan(
    String id,
    User user,

    Type type,
    LocalDate date,
    LocalTime time,
    String partyName,
    String title,
    String description,

    BigDecimal amount,
    BigDecimal remainingAmount,
    Currency currency,
    double interestRate,

    ModelStatus status,
    LocalDateTime createdAt,
    LocalDateTime updatedAt,
    String createdBy,
    String updatedBy) {

  public static enum Type {
    BORROW,
    LEND;

    public static Type from(String type) {
      if (type == null) {
        return null;
      }
      try {
        return valueOf(type);
      } catch (IllegalArgumentException e) {
        throw new IllegalArgumentException("Invalid loan type: " + type);
      }
    }
  }
}

File: mobile/app/lib/features/loan/domain/loan.dart

import 'package:bitapp/common/data/model/object_status.dart';
import 'package:bitapp/features/loan/domain/loan_type.dart';
import 'package:flutter/material.dart';

class Loan  {
  final String? id;
  final String title;
  final double amount;
  final DateTime date;
  final TimeOfDay time;
  final String? partyName;
  final String? description;
  final LoanType type;
  final String userId;
  final double? remainingAmount;
  final ObjectStatus? status;

  Loan({
    this.id,
    required this.title,
    this.description,
    required this.amount,
    required this.partyName,
    required this.date,
    required this.time,
    required this.type,
    required this.userId,
    this.remainingAmount,
    this.status,
  });
}
